cdap-hbase-compat-base/src/main/java/co/cask/cdap/replication/TableUpdater.java

```diff
@@ -91,9 +91,16 @@ public void run() {
             createTableIfNotExists(conf);
             tableExists = true;
           }
-          if (!cachedUpdates.isEmpty()) {
-            LOG.debug("Update Replication State table now. {} entries.", cachedUpdates.size());
-            writeState(cachedUpdates);
+          HashMap<String, Long> cachedUpdatesCopy;
+          synchronized (cachedUpdates) {
+            // Make a copy so HBase write can happen outside of the synchronized block
+            cachedUpdatesCopy = (HashMap<String, Long>) cachedUpdates.clone();
```

Comment:
Yes, we need that to avoid any updates between the copy and clear. The timer run is always in the same thread, the synchronized block here is only required to synchronize the copy+clear against updates. 
Also, since this only happens once every period [20s by default] and is only a memory copy operation by, the performance impact is negligible.